src/app/content/hooks/locationChange.ts

```diff
@@ -1,60 +1,113 @@
 import css from 'cnx-recipes/styles/output/intro-business.json';
-import { RouteHookBody } from '../../navigation/types';
+import { BOOKS } from '../../../config';
+import { Match, RouteHookBody } from '../../navigation/types';
 import { AppServices, FirstArgumentType, MiddlewareAPI } from '../../types';
+import { assertDefined } from '../../utils';
 import { receiveBook, receivePage, requestBook, requestPage } from '../actions';
 import { content } from '../routes';
 import * as select from '../selectors';
 import { ArchivePage, Book } from '../types';
-import { flattenArchiveTree, getContentPageReferences } from '../utils';
+import { flattenArchiveTree, getContentPageReferences, getPageIdFromUrlParam, getUrlParamForPageId } from '../utils';
 
 const fontMatches = css.match(/"(https:\/\/fonts\.googleapis\.com\/css\?family=.*?)"/);
 const fonts = fontMatches ? fontMatches.slice(1) : [];
 
 const hookBody: RouteHookBody<typeof content> = (services) => async({match}) => {
-  const {dispatch, getState, fontCollector, archiveLoader} = services;
+  const {fontCollector} = services;
+
+  fonts.forEach((font) => fontCollector.add(font));
+
+  await resolveBook(services, match).then(resolvePage(services, match));
+};
+
+const resolveBook = async(
+  services: AppServices & MiddlewareAPI,
+  match: Match<typeof content>
+): Promise<[Book, ReturnType<AppServices['archiveLoader']['book']>]> => {
+  const {dispatch, getState, archiveLoader} = services;
+  const [bookSlug, bookId, bookVersion] = await resolveBookReference(services, match);
+
+  const loader = archiveLoader.book(bookId, bookVersion);
   const state = getState();
-  const {bookId, pageId} = match.params;
   const bookState = select.book(state);
-  const pageState = select.page(state);
-  const book = bookState && bookState.shortId === bookId ? bookState : undefined;
-  const page = pageState && pageState.shortId === pageId ? pageState : undefined;
-  const promises: Array<Promise<any>> = [];
+  const book = bookState && bookState.id === bookId ? bookState : undefined;
 
-  fonts.forEach((font) => fontCollector.add(font));
+  if (book) {
+    return [book, loader];
+  }
+
+  const response = async(): Promise<[Book, ReturnType<AppServices['archiveLoader']['book']>]> => {
+    const newBook = {
+      ...await loader.load(),
+      slug: bookSlug,
+    };
+    return [newBook, archiveLoader.book(newBook.id, newBook.version)];
+  };
 
-  const [bookRefId, bookRefVersion, pageRefId] = match.state
-    ? [match.state.bookUid, match.state.bookVersion, match.state.pageUid]
-    : [bookId, undefined, pageId];
+  if (bookSlug !== select.loadingBook(state)) {
+    dispatch(requestBook(bookSlug));
+    return await response().then((params) => {
```

Comment:
Why combine both `await` and Promise syntax?

```suggestion
    const params = await response();
    dispatch(receiveBook(params[0]));
    return params;
```